### ISSUE TITLE
Wrrong match with nested !()

### ISSUE THREAD

**Issue by rockandska** [facelessuser/wcmatch#29]
Hi there,

Looking for a library to match extglob pattern against stdin, i've give a test to your library and write a little script to test the differences between wcmatch and bash extglob and found at least one difference.

Pattern: '!(*.@(so.!(conf|*/*).test))'
Difference:
```
Extglob pattern: !(*.@(so.!(conf|*/*).test))

+++ test.so.conf.test						+++ test.so.conf.test
--- test.so.conf.conf.test					--- test.so.conf.conf.test
--- test.so.confe.test						--- test.so.confe.test
--- test.so.econf.test						--- test.so.econf.test
+++ conf.eps.pdf						+++ conf.eps.pdf
--- conf.so.ps.test						--- conf.so.ps.test
+++ so.test.conf.ps						+++ so.test.conf.ps
+++ test.so.test						+++ test.so.test
+++ test.so.d/.test					      |	--- test.so.d/.test
+++ test.so./ff/.test					      |	--- test.so./ff/.test
```

Please find my little script [here](https://github.com/rockandska/repro-issues/tree/master/facelessuser/wcmatch/issue_29) just in case i've missed something.

you could use your own list of file if you would like:
```
$ echo "test.so.d/.test" | ./compare_wcmatch_extglob.sh  '!(*.@(so.!(conf|*/*).test))'

Extglob pattern: !(*.@(so.!(conf|*/*).test))

+++ test.so.d/.test					      |	--- test.so.d/.test
```

Not relate, any chance to have a compatibility with python 2.x ?

Regards,

**Comment by rockandska** [comment]
Another example:
```
$  ./compare_wcmatch_extglob.sh '!(*.@(?(e)ps|pdf))'

Extglob pattern: !(*.@(?(e)ps|pdf))

+++ test.so.conf.test						+++ test.so.conf.test
+++ test.so.conf.conf.test					+++ test.so.conf.conf.test
+++ test.so.confe.test						+++ test.so.confe.test
+++ test.so.econf.test						+++ test.so.econf.test
--- conf.eps.pdf						--- conf.eps.pdf
+++ conf.so.ps.test						+++ conf.so.ps.test
--- so.test.conf.ps						--- so.test.conf.ps
+++ test.so.test						+++ test.so.test
+++ test.so.d/.test					      |	--- test.so.d/.test
+++ test.so./ff/.test					      |	--- test.so./ff/.test
```

**Comment by facelessuser** [comment]
I'm not sure exactly how you are testing, but here I create a directory with the following files and folders:

```
test.so.conf.test
test.so.conf.conf.test
test.so.confe.test
test.so.econf.test
conf.eps.pdf
conf.so.ps.test
so.test.conf.ps
test.so.test
test.so.d/.test
test.so./ff/.test
```

Bash:

```
$ echo !(*.@(so.!(conf|*/*).test))
conf.eps.pdf so.test.conf.ps test.so. test.so.conf.test test.so.d test.so.test
```

Wcmatch Glob

```py
>>> from wcmatch import glob
>>> glob.glob('!(*.@(so.!(conf|*/*).test))', flags=glob.E)
['conf.eps.pdf', 'so.test.conf.ps', 'test.so.', 'test.so.conf.test', 'test.so.d', 'test.so.test']
```

Looks like its working to me.

**Comment by facelessuser** [comment]
Second test also looks fine:

```
$ echo !(*.@(?(e)ps|pdf))
conf.so.ps.test test.so. test.so.conf.conf.test test.so.conf.test test.so.confe.test test.so.d test.so.econf.test test.so.test
```

```py
>>> from wcmatch import glob
>>> glob.glob('!(*.@(?(e)ps|pdf))', flags=glob.E)
['conf.so.ps.test', 'test.so.', 'test.so.conf.conf.test', 'test.so.conf.test', 'test.so.confe.test', 'test.so.d', 'test.so.econf.test', 'test.so.test']
```

I feel like you are somehow testing something different?

**Comment by facelessuser** [comment]
As far as Python 2 support is concerned, I have no plans to support it. Python 2 is officially dead (end of life) come 2020. The effort required to get Python 2 support in is simply not worth it to me as I don't use this library in Python 2. People are free to back port it if they like.

**Comment by rockandska** [comment]
> I'm not sure exactly how you are testing, but here I create a directory with the following files and folders:

Link was added on the first post:
https://raw.githubusercontent.com/rockandska/repro-issues/master/facelessuser/wcmatch/issue_29/compare_wcmatch_extglob.sh

To resume i test like this with bash:  
```bash
$ [[ "test.so./ff/.test" == !(*.@(so.!(conf|*/*).test)) ]] && echo "match" || echo "not match"
match
$
```
And like this with python:
```python
>>> from wcmatch import glob
>>> glob.globmatch('test.so./ff/.test',r'!(*.@(so.!(conf|*/*).test))', flags=glob.EXTGLOB)
False
```

**Comment by facelessuser** [comment]
You are testing something very different, and I'll explain.

You'll notice that the way I tested things in my initial post, calling `echo` in bash on a file structure returned files with `glob` logic and did not match what you are doing.

In bash, you are testing things on strings which is simulating `fnmatch` (file name match) logic. The `/` are not treated as directory separators. So that is why it matches for you. `wcmatch.globmatch` matches what `glob` globs, meaning it will treat `/` as directory separators and match it as if it is looking at a file path and not a file name. It will break up the path accordingly.

To get what your bash script is doing, which is not a proper simulation of `glob`, you can use `fnmatch`:

```
$ [[ "test.so./ff/.test" == !(*.@(so.!(conf|*/*).test)) ]] && echo "match" || echo "not match" match
match
```

```
>>> fnmatch.fnmatch('test.so./ff/.test', '!(*.@(so.!(conf|*/*).test))', flags=fnmatch.E)
True
```

Basically you are comparing file name match logic with glob logic which are not the same. When you compare `fnmatch` to `fnmatch`, it matches (see above). And when you compare `glob` logic to `glob` logic, it matches as well, which I show in previous posts.

This is a test issue, not a `wcmatch` issue.

**Comment by facelessuser** [comment]
Just to illustrate a bit further as this issue was originally complaining about `globmatch`, this example will show that `globmatch` will return everything that `glob` matches. We will use `globfilter` which is just `globmatch` for a list of files:

```py
>>> from wcmatch import glob
>>> glob.glob('!(*.@(so.!(conf|*/*).test))', flags=glob.E)                                                              
['conf.eps.pdf', 'so.test.conf.ps', 'test.so.', 'test.so.conf.test', 'test.so.d', 'test.so.test']
>>> files = glob.glob('!(*.@(so.!(conf|*/*).test))', flags=glob.E)
>>> glob.globfilter(files, '!(*.@(so.!(conf|*/*).test))', flags=glob.E)                                                 
['conf.eps.pdf', 'so.test.conf.ps', 'test.so.', 'test.so.conf.test', 'test.so.d', 'test.so.test'] 
```

Everything `glob` found on an actual file system is then matched with `globfilter` (`globmatch` for a list of names) and returned. `test.so./ff/.test` was not matched by `glob`, so `globmatch` will not match it either:

```py
>>> glob.globmatch('test.so./ff/.test', '!(*.@(so.!(conf|*/*).test))', flags=glob.E)
False
```

I think with this, the issue can be marked as working as expected and closed.

**Comment by facelessuser** [comment]
Feel free to reopen if I have somehow misunderstood something, but based on my current analysis, `globmatch` seems to be working as intended.

**Comment by rockandska** [comment]
Thanks for all those explanations, i didn't get at first the slightly difference between `glob.globmatch` / `fnmatch.fnmatch` and the fact my test case was testing some different things.

Thanks for this library too, i think i will use it instead of continue to looking for the golden regex able to mimic the effect of `!()` even when nested since i've spent already many hours on it without success. Good job !

**Comment by facelessuser** [comment]
@rockandska, if in the future you find a concrete example where `!()` appears to behave different in a `glob` environment, please do file an issue, as I'll be more than happy to fix any erroneous behavior.

`wcmatch` has implemented both `fnmatch` and `glob`. The only difference between the two is that `glob` will handle file separators differently, and optionally has feature that enables `**` which spans directory separators. In `glob`, directory separators will will never be matched by `*`, and `!()` are defined by filename boundaries. Based on your tests, it seems like you may want `wcmatch.fnmatch` as it appears you want to treat the entire file path as a filename. If not, you can use `wcmatch.globmatch` will treat file separators characters as actual file name separators.

I think what you are experiencing is that bash can tell when it is looking at a string or directly at the file system. When you are directly accessing the file system, I believe it is only ever comparing the file and folder names separately as it is physically crawling the file system. When globbing, I don't think bash ever even sees the `/` as that isn't how the names are given to it. In normal cases, `/` is an illegal file/folder name character. When on a real file system, bash never has to check for `/` as it will never actually read the file system and return a name with `/`.

 When you just pipe a string, it assumes that it got a full filename since it isn't actively crawling the filesystem, and it evaluates it as a filename, meaning the `/` has no significance in that case. That is why if you use `wcmatch.fnmatch` it works the way **you** are testing it. And that is also why when **I** am testing against the actual file system in bash, it doesn't work like your test, and models `wcmatch.globmatch`.

I can see how this confusion can come about.
